### PR TITLE
feat(dynamic-forms): adds additional meta info for dynamic forms

### PR DESCRIPTION
**Pull request by mario-subo**
dynamic forms hook now returns meta info for the form with info on visibility, validity and error
state

### CODE REVIEW

File: lib/src/hooks/useDynamicForm.tsx [hunk 1]
```diff
@@ -254,11 +264,57 @@ export function useDynamicForm(
...
+                        break;
+                    case "Checkbox":
+                        isValid = !!itemState;
+                        break;
+                    case "Stepper":
+                        isValid = Number.isInteger(itemState);
+                        break;
+                    default:
+                        isValid = null;
+                        console.warn(`Could not determine is state is valid for control type ${controlType}`);
```
Comment (omarVodiaK) [inline lib/src/hooks/useDynamicForm.tsx:298]:
```suggestion
                        console.warn(`Could not determine if state is valid for control type ${controlType}`);
```

Comment (mario-subo) [reply lib/src/hooks/useDynamicForm.tsx]: thanks

File: lib/src/hooks/useDynamicForm.tsx [hunk 2]
```diff
@@ -254,11 +264,57 @@ export function useDynamicForm(
...
         [state]
     );
+
+    const meta = React.useMemo(() => {
+        let newMeta: DynamicFormMetaData = {};
+
+        sections?.forEach(({ key: sectionKey, items }) => {
+            newMeta[sectionKey] = {};
+            items?.forEach(({ key, controlType }) => {
+                const itemState: DynamicFormInternalStateValue = state && state[sectionKey] && state[sectionKey][key];
```
Comment (eweseong) [inline lib/src/hooks/useDynamicForm.tsx:274]:
```
const itemState: DynamicFormInternalStateValue = state && state[sectionKey] && state[sectionKey][key];
```
from the usage, itemState does not seemed to be intended as boolean value, perhaps the intention was to use `||`?

Comment (mario-subo) [reply lib/src/hooks/useDynamicForm.tsx:274]: Yeah the intention here is to have a falsy value if the value does not exist and the actual value if it does.

**Review by eweseong — APPROVED** [review]

File: lib/src/hooks/useDynamicForm.tsx [hunk 3]
```diff
@@ -145,6 +158,7 @@ export function useDynamicForm(
     });
     const [state, setState] = useState<DynamicFormInternalState>(initialState);
     const [errorMessages, setErrorMessages] = useState<DynamicFormErrors>({});
+    const [touched, setTouched] = useState<boolean>(false);
```
Comment (eweseong) [inline lib/src/hooks/useDynamicForm.tsx:161]: maybe can refactor to the more commonly use term `dirty`?

**Review by eweseong — APPROVED** [review]

File: lib/src/hooks/useDynamicForm.tsx (hunk 3 above)
Comment (mario-subo) [reply lib/src/hooks/useDynamicForm.tsx:161]: Ah yes dirty is more accurate! thanks!

File: docs/src/pages/docs/dynamic-forms.tsx [hunk 4]
```diff
@@ -884,43 +589,26 @@ const ShowMeEverything: React.FC = () => {
...
     ];
 
-    const [renderForm, state, , setErrors] = useDynamicForm(sections);
+    const [renderForm, state, , setErrors, meta, touched] = useDynamicForm(sections);
+
+    console.log("touched: ");
+    console.log(touched);
+    console.log("META: ");
+    console.log(JSON.stringify(meta, null, 4));
+    console.log("--------");
```
Comment (eweseong) [inline docs/src/pages/docs/dynamic-forms.tsx:603]: are these testing code? let's remove them if they are only meant for testing

Comment (mario-subo) [reply docs/src/pages/docs/dynamic-forms.tsx:603]: omg thanks I totally missed that

Comment (mario-subo) [reply docs/src/pages/docs/dynamic-forms.tsx]: looks like I ended up pushing the wrong file

**Review by kherP — APPROVED** [review]

**Review by eweseong — APPROVED** [review]

**Review by omarVodiaK — APPROVED** [review]
should add some unit tests for this

**Pull request merged**